### PR TITLE
Forced one time keys

### PR DESCRIPTION
**Pull request by davidrusu**
- [x] spendbook is keyed by owner
- [x] dbc name is it's owner
- [x] Implement the wellknown / blinded / spend key pattern

### CODE REVIEW

**Comment by dan-da** [comment]
I see some changes in mint-repl.  Did you test that it still works, in particular the commands:
1. reissue_ez and
2. prepare_tx, sign_tx, prepare_reissue, reissue

**Comment by dan-da** [comment]
I would like to see a comment somewhere, maybe in lib.rs that gives an overview of forced-one-time-keys and how it works, to provide context for the individual changes/pieces.

File: src/dbc.rs [hunk 1]
```diff
@@ -7,22 +7,45 @@
...
     pub content: DbcContent,
     pub transaction: DbcTransaction,
-    pub transaction_sigs: BTreeMap<DbcContentHash, (PublicKey, Signature)>,
+    pub transaction_sigs: BTreeMap<SpendKey, (PublicKey, Signature)>,
 }
 
 impl Dbc {
-    pub fn name(&self) -> Hash {
-        self.content.hash()
+    pub fn spend_key(&self) -> SpendKey {
```
Comment (dan-da) [inline src/dbc.rs:25]: We should doc-comment all public APIs.

File: src/lib.rs [hunk 2]
```diff
@@ -27,19 +21,23 @@ mod mint;
...
     },
     mint::{
-        Mint, MintNodeSignatures, ReissueRequest, ReissueShare, ReissueTransaction,
-        SimpleSpendBook, SpendBook, GENESIS_DBC_INPUT,
+        genesis_dbc_input, Mint, MintNodeSignatures, ReissueRequest, ReissueShare,
+        ReissueTransaction, SimpleSpendBook, SpendBook,
     },
 };
 
+/// These typdefs are to simplify algorithm for now and will be removed for production.
```
Comment (dan-da) [inline src/lib.rs:37]: is this comment still accurate?

File: src/lib.rs [hunk 3]
```diff
@@ -72,14 +61,31 @@ impl AsRef<[u8]> for Hash {
 #[cfg(test)]
 use rand::distributions::{Distribution, Standard};
 
+#[derive(Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct SpendKey(pub PublicKey);
+
+// Display Hash value as hex in Debug output.  consolidates 36 lines to 3 for pretty output
+// and the hex value is the same as sn_dbc_mint display of DBC IDs.
+impl fmt::Debug for SpendKey {
```
Comment (dan-da) [inline src/lib.rs:69]: maybe should impl fmt::Debug for Hash also?

**Review by dan-da — COMMENTED** [review]
lgtm.   would like to see more comments, per above.

File: src/dbc.rs (hunk 1 above)
Comment (davidrusu) [reply src/dbc.rs:25]: Yes we should, I'll add a few here for now but we should add the `# deny(missing docs)` lint once the library is stable.

File: src/lib.rs (hunk 2 above)
Comment (davidrusu) [reply src/lib.rs:37]: Probably not, I've removed it.

File: src/lib.rs (hunk 3 above)
Comment (davidrusu) [reply src/lib.rs:69]: Yep, we have it already implemented

**Comment by davidrusu** [comment]
@dan-da I added some docs on one-time-keys and regenerated the mint-repl sample runs, please take another look

File: examples/mint-repl/sample_runs/reissue_ez.txt [hunk 4]
```diff
@@ -27,62 +27,59 @@ Number of Mint Nodes: 1
...
...7b3d1d4c91497caac22cc836534597cc58cef6743ed4d5724756d1b22d95c7ab4b8d11c0785f60b34bce91afb47f773aba0fa4c67bbd0cde38baaf761ed2d301b83ac8a07c9641fe7c3745e96e8f273bb0be2c8ff458e2440658fcb14c3e20768df414712535e5dc392012e58aee62f4721712a9b0e6692b02ab7e1f8e5c80a158577d332ecb8190ec20aaa401d6da42286c4d7eb7fc38c60f059ed05f1c705066bbe02f0486c686362c429d99e534e7dcc341a789c70cae28152dacf0b24271b4b4cc90b8c0ace262c09725b4fd626d3a5158f7df8fd3ee6162a65c700051851b125403f416029b955e5bc770b6f7d119f7273460bd70a24169d7db70055430481dbcf2cf9c1c6a390df44be21fa58cde20b48b135c215845fc590105babb7f60668753de793d3bed1c12abcccb5b218022357aeba4785a00000000000002a04d8ea7ca1e65de3f6c5bcc481c6fbe78e77b28f6fc255604d18b4c7f9b3b05bc32ce433590953a49e0d7fb64ab3f592ac41d25a5802ef0bbf517a1e384652c002002015b7fdc89f960743eb645076c0c23845e3135bd18e47dee60c23ef647917d5e277ba4a0c002dd80436aa7ebc80fb55864967a2f37a1fdb6fbae737f53921ad8d7da966423d88d68c34388fa265b1470ae1fe9a2d69767eb193b3ff6074d9d638bc210f25b1700000000000000287634ac86d6ee1afbac7b3ea56c6b281945562b035f958efdbc645f8213fe9a49df0c54d127c767ecbbe7f0dcfa415ab7bbc622db0af03afbef1a7af93fe8556c58ac1b173f3a4ea105b974974f8c68c30faca94f8c63952694d79731a7d3f1970000000000000001
```
Comment (dan-da) [inline examples/mint-repl/sample_runs/reissue_ez.txt:63]: total dbc size increases a bit, not hugely though.

File: examples/mint-repl/sample_runs/reissue_ez.txt [hunk 5]
```diff
@@ -27,62 +27,59 @@ Number of Mint Nodes: 1
...
 
 
 -- SpendBook --
 
-  0000000000000000000000000000000000000000000000000000000000000000
-
+  97f1d3a73197d7942695638c4fa9ac0fc3688c4f9774b905a14e3a3f171bac586c55e83ff97a1aeffb3af00adb22c6bb
```
Comment (dan-da) [inline examples/mint-repl/sample_runs/reissue_ez.txt:68]: Would it make any sense to hash the key so it's smaller, eg as spendbook key?   Or would that break key derivation somehow?

File: src/spend_book.rs [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+// # SpendBook and Forced One-Time-Keys
+//
+// The SpendBook logs each spent DBC along with the reissue transaction it was spent in.
```
Comment (dan-da) [inline src/spend_book.rs:11]: We should discuss about this.

Let's say we have a reissue with 20 inputs and 20 outputs.    iiuc, the spendbook is logging the same transaction for each DBC, ie 40 times.   That's a lot of duplication.

Or maybe not, if data store (merkle reg?) keeps just one copy ot the transaction "value" with 40 pointers to it?

**Review by dan-da — APPROVED** [review]
lgtm

File: examples/mint-repl/sample_runs/reissue_ez.txt (hunk 5 above)
Comment (davidrusu) [reply examples/mint-repl/sample_runs/reissue_ez.txt:68]: Yeah, I think we can safely do that, but I'd like to keep things unobscured, at least the first pass

File: src/spend_book.rs (hunk 6 above)
Comment (davidrusu) [reply src/spend_book.rs:11]: As discussed in our HO, likely an area for future optimization, we can leave this as an implementation detail of the spendbook, for sn_dbc we just need to maintain this API of spend key -> transaction.

File: examples/mint-repl/sample_runs/reissue_ez.txt [hunk 7]
```diff
@@ -27,62 +27,59 @@ Number of Mint Nodes: 1
...
+     blinding_factor: 02f0b9816e3b7731e05902d3cd0e1f89b9faed861da3e16475bb15dcd3237716
 
-output_number: 0
-
-owner: 47c96e1397a0d7b26783624f14fc0b7dde6eb475e7fc0f5caa58d77bffd96bfa107a7f9006cd2b2c0062dc2ab943c0b00000000000000001
+owner: e13050c7dd443e07e515eb2b901460da291447eb3bd6773c5289d369fb7f36c6b3b932f5340a290bc0f3d5c248adad840000000000000001
 
 inputs:
-  0000000000000000000000000000000000000000000000000000000000000000
+  97f1d3a73197d7942695638c4fa9ac0fc3688c4f9774b905a14e3a3f171bac586c55e83ff97a1aeffb3af00adb22c6bb
```
[the remark this answers is not in the document]
Comment (davidrusu) [reply examples/mint-repl/sample_runs/reissue_ez.txt:57]: yeah the BLS pub keys are a bit larger than a 256bit hash

File: examples/mint-repl/sample_runs/reissue_ez.txt (hunk 4 above)
Comment (davidrusu) [reply examples/mint-repl/sample_runs/reissue_ez.txt:63]: likely due to those BLS keys replacing the dbc hashes

**Pull request merged**